### PR TITLE
Change name space level

### LINKED ISSUE #36: Change the namespace level
Currently, the name-space is on the YAML children level.  

Per discussion with @alfsuse we decided to move the level up, so one plugin is specific for one name-space. 
If the user needs more plugins, it will use the #12 

This change will make the plugins easier to understand and to write.

### PR DESCRIPTION
**Pull request by Gsantomaggio**
Closes #36 

The right plugin to test it is:

```
k3ai-cli apply kubeflow-pipelines
```

We also changed the[ `deploy delete` ](https://github.com/kf5i/k3ai-core/pull/37/files#diff-9979203cc351d5261fa3c0e542f200d2a5f417173473a74b6df77a013ffb470bR45) loop order, the resources should be deleted in the inverse order of their creation.

Added some sleep during the operations.

### CODE REVIEW

File: internal/k8s/kctl/deploy.go [hunk 1]
```diff
@@ -19,12 +21,23 @@ type Wait interface {
 	Process(labels []string)
 }
 
+func pause() {
```
Comment (harsimranmaan) [inline internal/k8s/kctl/deploy.go:24]: Were there any race conditions that you ran into? We can implement wait to handle these scenarios too. I haven't run into issues so far.

File: internal/plugins/plugin_spec_test.go [hunk 2]
```diff
@@ -6,14 +6,19 @@ import (
 	"testing"
 )
 
+func (receiver YamlSpec) name() {
```
Comment (Gsantomaggio) [inline internal/plugins/plugin_spec_test.go:9]: have to remote this

File: internal/k8s/kctl/deploy.go (hunk 1 above)
Comment (Gsantomaggio) [reply internal/k8s/kctl/deploy.go:24]: I and @alfsuse had some problem with some plugin in k3ai, we put a pause and we solved. 
This is why you see the [pauses](https://github.com/kf5i/k3ai-plugins/blob/main/plugin_jupyter-scipy/install#L114)  on the plugins. (We haven't investigated in detail btw)

Comment (harsimranmaan) [reply internal/k8s/kctl/deploy.go:24]: ok. I am cool with this for now. We should revisit this when we implement wait.

File: internal/k8s/kctl/deploy.go [hunk 3]
```diff
@@ -33,7 +46,18 @@ func Apply(config Config, plugin plugins.PluginSpec, evt Wait) error {
 
 // Delete removes the plugin from the cluster
 func Delete(config Config, plugin plugins.PluginSpec) error {
-	return handleYaml(config, delete, plugin)
+	for i := len(plugin.Yaml) - 1; i >= 0; i-- {
+		yamlSpec := plugin.Yaml[i]
+		err := execute(config, k3sExec, kubectl, delete,
+			decodeType(yamlSpec.Type), yamlSpec.URL, "-n", plugin.NameSpace)
+		if err != nil {
+			log.Fatalf("Error during delete: %s", err.Error())
```
Comment (harsimranmaan) [inline internal/k8s/kctl/deploy.go:54]: return errors.Wrap() here.

File: internal/k8s/kctl/deploy.go [hunk 4]
```diff
@@ -33,7 +46,18 @@ func Apply(config Config, plugin plugins.PluginSpec, evt Wait) error {
...
+		}
+		pause()
+	}
+	_ = deleteNameSpace(config, plugin.NameSpace)
```
Comment (harsimranmaan) [inline internal/k8s/kctl/deploy.go:58]: On a more closer look, I think we should not delete namespaces. As there could be multiple plugins installed in a namespace.

File: internal/plugins/plugin_spec.go [hunk 5]
```diff
@@ -16,18 +16,18 @@ const (
...
-	URL       string `yaml:"url"`
-	NameSpace string `yaml:"namespace,omitempty"`
-	Type      string `yaml:"type,omitempty"`
+	URL  string `yaml:"url"`
+	Type string `yaml:"type,omitempty"`
 }
 
 //PluginSpec is the specification of each k3ai plugin
 type PluginSpec struct {
+	NameSpace  string     `yaml:"namespace,omitempty"`
```
Comment (harsimranmaan) [inline internal/plugins/plugin_spec.go:25]: nit: Can we use `Namespace` here?

File: internal/k8s/kctl/deploy.go (hunk 4 above)
Comment (Gsantomaggio) [reply internal/k8s/kctl/deploy.go:58]: good point. We could check if there are dependencies before remove it.

File: internal/k8s/kctl/deploy.go [hunk 6]
```diff
@@ -19,12 +21,23 @@ type Wait interface {
...
-	err := handleYaml(config, apply, plugin)
-	if err != nil {
-		return err
+	_ = createNameSpace(config, plugin.NameSpace)
+	pause()
+	for _, yamlSpec := range plugin.Yaml {
+		err := execute(config, k3sExec, kubectl, apply,
+			decodeType(yamlSpec.Type), yamlSpec.URL, "-n", plugin.NameSpace)
+		if err != nil {
+			log.Fatalf("Error during create: %s", err.Error())
```
Comment (harsimranmaan) [inline internal/k8s/kctl/deploy.go:36]: You can return the error with a Wrap instead of log.Fatal

Comment (harsimranmaan) [inline internal/k8s/kctl/deploy.go:36]: You can return this error to the caller

File: internal/k8s/kctl/deploy.go (hunk 4 above)
Comment (Gsantomaggio) [reply internal/k8s/kctl/deploy.go:58]: I will remove the `delete` I will open another issue to handle it.

File: internal/k8s/kctl/deploy.go (hunk 6 above)
Comment (Gsantomaggio) [reply internal/k8s/kctl/deploy.go:36]: I voluntary want that the cycle continues. There could be temporary problems ( especially for the kubeflow installation) there are a lot of retries. So I'd leave it as it then we will see.

File: internal/k8s/kctl/deploy.go (hunk 3 above)
Comment (Gsantomaggio) [reply internal/k8s/kctl/deploy.go:54]: same https://github.com/kf5i/k3ai-core/pull/37/files#r512535062

File: internal/k8s/kctl/deploy.go (hunk 6 above)
Comment (harsimranmaan) [reply internal/k8s/kctl/deploy.go:36]: Fatalf panics with error. You might want to println in that case

**Review by harsimranmaan — APPROVED** [review]
lgtm. Thanks

**Pull request merged**